server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -500,15 +503,21 @@ public void updateShardState(final ShardRouting newRouting,
                                  * numbers. To ensure that this is not the case, we restore the state of the local checkpoint tracker by
                                  * replaying the translog and marking any operations there are completed.
                                  */
-                                final Engine engine = getEngine();
-                                engine.restoreLocalCheckpointFromTranslog();
+                                if (seqNoStats().getLocalCheckpoint() < maxSeqNoOfResettingEngine) {
```

Comment:
having this `maxSeqNoOfResettingEngine` variable is messy. Can't we get the max sequence number information from the translog? We essentially want to check if Lucene needs to catch up with what's in the translog.